### PR TITLE
Fix windows bug: Backslashes causing issues with listing and downloading project contents

### PR DESCRIPTION
**Pull request by inaod568**
> **Before submitting the PR, please go through the sections below and fill in what you can. If there are any items that are irrelevant for the current PR, remove the row. If a relevant option is missing, please add it as an item and add a PR comment informing that the new option should be included into this template.**

> **All _relevant_ items should be ticked before the PR is merged**

# Description

Several users reported that there were issues with listing the project contents when the upload had been done from Windows. This PR solves that issue by changing instances of `os.path` etc to `pathlib` functionality.

- [x] Add a summary of the changes and the related issue
- [x] Add motivation and context regarding why the change is needed
- [x] Fixes DDS-1411

## Type of change

- [x] Bug fix (non-breaking)

# Checklist:

## General

- [x] [Changelog](../CHANGELOG.md): New row added. Not needed when PR includes _only_ tests.
- [x] Code change
  - [x] Self-review of code done
  - [x] Comments added, particularly in hard-to-understand areas
  - [x] Documentation is updated

## Repository / Releases

- [x] Rebase / update of branch done

## Checks

- [ ] Formatting: Black & Prettier checks pass
- [ ] CodeQL passes
- [x] Tests
  - [x] The tests pass

### CODE REVIEW

File: dds_cli/__main__.py [hunk 1]
```diff
@@ -304,12 +305,14 @@ def list_projects_and_contents(
                                 # If didn't enter anything, convert to None and exit
                                 except (KeyboardInterrupt, AssertionError):
                                     break
-
+                                
                                 # Prepend existing file path
                                 if last_folder is not None and folder is not None:
-                                    folder = os.path.join(last_folder, folder)
+                                    # folder = str(pathlib.Path(last_folder, folder).as_posix())
```
Comment (inaod568) [inline dds_cli/__main__.py:311]:
```suggestion
```

File: dds_cli/__main__.py [hunk 2]
```diff
@@ -304,12 +305,14 @@ def list_projects_and_contents(
...
+                                    folder = pathlib.Path(last_folder, folder)
 
                                 # List files
+                                LOG.info(folder)
```
Comment (inaod568) [inline dds_cli/__main__.py:315]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 3]
```diff
@@ -136,6 +137,8 @@ def list_files(self, folder: str = None, show_size: bool = False):
             error_message="Failed to get list of files in project",
             timeout=DDSEndpoint.TIMEOUT,
         )
+        LOG.debug(response)
+
```
Comment (inaod568) [inline dds_cli/data_lister.py:141]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 4]
```diff
@@ -227,6 +230,7 @@ class FileTree:
             name: str = None
 
         def __api_call_list_files(folder: str):
+            LOG.debug(f"Folder: {folder}")
```
Comment (inaod568) [inline dds_cli/data_lister.py:233]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 5]
```diff
@@ -236,6 +240,8 @@ def __api_call_list_files(folder: str):
                 headers=self.token,
                 error_message="Failed to list the project's directory tree",
             )
+            
+            LOG.debug(f"project contents: {resp_json}")
```
Comment (inaod568) [inline dds_cli/data_lister.py:244]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 6]
```diff
@@ -283,10 +290,17 @@ def __construct_file_tree(folder: str, basename: str) -> Tuple[FileTree, int, in
                 if not is_folder:
                     tree.subtrees.append((escape(f["name"]), f.get("size") if show_size else None))
                 else:
+                    # TODO: JOIN IS MESSING SHIT UP
+                    LOG.debug("")
+                    LOG.debug(f'join: {os.path.join(folder, f["name"]) if folder else f["name"]}')
+                    LOG.debug(f'pathlib: {pathlib.Path(folder, f["name"]).as_posix() if folder else f["name"]}')
+                    LOG.debug("")
```
Comment (inaod568) [inline dds_cli/data_lister.py:297]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 7]
```diff
@@ -283,10 +290,17 @@ def __construct_file_tree(folder: str, basename: str) -> Tuple[FileTree, int, in
...
                     subtree, _max_string, _max_size = __construct_file_tree(
-                        os.path.join(folder, f["name"]) if folder else f["name"],
+                        pathlib.Path(folder, f["name"]).as_posix() if folder else f["name"],
+                        # os.path.join(folder, f["name"]) if folder else f["name"],
```
Comment (inaod568) [inline dds_cli/data_lister.py:300]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 8]
```diff
@@ -283,10 +290,17 @@ def __construct_file_tree(folder: str, basename: str) -> Tuple[FileTree, int, in
...
                         f"[bold deep_sky_blue3]{escape(f['name'])}",
                     )
+                    LOG.debug(f"subtree: {subtree}\t_max_string: {_max_string}\t_max_size: {_max_size}")
```
Comment (inaod568) [inline dds_cli/data_lister.py:303]:
```suggestion
```

File: dds_cli/data_lister.py [hunk 9]
```diff
@@ -323,7 +337,8 @@ def __construct_file_dict_tree(folder: str) -> dict:
                         tree[f["name"]]["size"] = f.get("size")
                 else:
                     children = __construct_file_dict_tree(
-                        os.path.join(folder, name) if folder else name
+                        pathlib.Path(folder, name).as_posix() if folder else name,
+                        #os.path.join(folder, name) if folder else name
```
Comment (inaod568) [inline dds_cli/data_lister.py:341]:
```suggestion
```

File: dds_cli/data_putter.py [hunk 10]
```diff
@@ -221,6 +221,7 @@ def __init__(
             wait_task = progress.add_task("Collecting and preparing data", step="prepare")
 
             # Get file info
+            LOG.debug(f"2: {[type(x) for x in source]}")
```
Comment (inaod568) [inline dds_cli/data_putter.py:224]:
```suggestion
```

File: dds_cli/data_putter.py [hunk 11]
```diff
@@ -233,6 +234,7 @@ def __init__(
 
             # Check which, if any, files exist in the db
             files_in_db = self.filehandler.check_previous_upload(token=self.token)
+            LOG.debug(f"files in db: {files_in_db}")
```
Comment (inaod568) [inline dds_cli/data_putter.py:237]:
```suggestion
```

File: dds_cli/data_putter.py [hunk 12]
```diff
@@ -265,6 +267,8 @@ def __init__(
     @subpath_required
     def protect_and_upload(self, file, progress):
         """Process and upload the file while handling the progress bars."""
+        LOG.debug(f"file: {file}")
+
```
Comment (inaod568) [inline dds_cli/data_putter.py:271]:
```suggestion
```

File: dds_cli/data_putter.py [hunk 13]
```diff
@@ -305,6 +309,7 @@ def protect_and_upload(self, file, progress):
             LOG.debug(
                 f"File successfully encrypted: {escape(file)}. New location: {escape(str(file_info['path_processed']))}"
             )
+
```
Comment (inaod568) [inline dds_cli/data_putter.py:312]:
```suggestion
```

File: dds_cli/data_putter.py [hunk 14]
```diff
@@ -392,17 +397,19 @@ def add_file_db(self, file):
         # Get file info and specify info required in db
         fileinfo = self.filehandler.data[file]
         params = {"project": self.project}
+
```
Comment (inaod568) [inline dds_cli/data_putter.py:400]:
```suggestion

```

File: dds_cli/data_putter.py [hunk 15]
```diff
@@ -392,17 +397,19 @@ def add_file_db(self, file):
...
+            "name_in_bucket": fileinfo["path_remote"],
+            "subpath": fileinfo["subpath"],
             "size": fileinfo["size_raw"],
             "size_processed": fileinfo["size_processed"],
             "compressed": not fileinfo["compressed"],
             "salt": fileinfo["salt"],
             "public_key": fileinfo["public_key"],
             "checksum": fileinfo["checksum"],
         }
+        LOG.debug(f"file_info to be added to db: \n{file_info}")
```
Comment (inaod568) [inline dds_cli/data_putter.py:412]:
```suggestion
```

File: dds_cli/file_handler.py [hunk 16]
```diff
@@ -40,19 +40,19 @@ def __init__(self, user_input, local_destination, project=None):
         self.project = project
         self.local_destination = local_destination
         self.data_list = []
-        if source is not None:
+        if source:
             self.data_list += list(source)
-        if source_path_file is not None:
-            source_path_file = pathlib.Path(source_path_file)
+        if source_path_file:
+            # source_path_file = pathlib.Path(source_path_file)
```
Comment (inaod568) [inline dds_cli/file_handler.py:46]:
```suggestion
```

File: dds_cli/file_handler.py [hunk 17]
```diff
@@ -40,19 +40,19 @@ def __init__(self, user_input, local_destination, project=None):
...
                 try:
                     with source_path_file.resolve().open(mode="r") as spf:
-                        self.data_list += spf.read().splitlines()
+                        self.data_list += [pathlib.Path(x) for x in spf.read().splitlines()]
                 except OSError as err:
                     raise dds_cli.exceptions.UploadError(
                         f"Failed to get files from source-path-file option: {err}"
                     )
-
+        LOG.debug(f"3: {self.data_list}")
```
Comment (inaod568) [inline dds_cli/file_handler.py:55]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 18]
```diff
@@ -48,9 +48,10 @@ def __init__(self, user_input, temporary_destination, project, remote_destinatio
 
         # Remove duplicates and save all files for later use
         all_files = set(self.data_list)
-
+        LOG.debug(f"4: {[type(x) for x in all_files]}")
+
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:52]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 19]
```diff
@@ -62,9 +63,12 @@ def __init__(self, user_input, temporary_destination, project, remote_destinatio
             )
 
         # Get absolute paths for all data
+        # os.path.expanduser(path): e.g. C:\Users\inaod568/repos/dds_cli
+        # path.expanduser(), pathlib.Path: e.g. C:\Users\inaod568\repos\dds_cli 
         self.data_list = [
-            pathlib.Path(os.path.abspath(os.path.expanduser(path))) for path in self.data_list
+            pathlib.Path(os.path.abspath(path.expanduser())) for path in self.data_list
         ]
+        LOG.debug(f"os: {self.data_list}")
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:71]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 20]
```diff
@@ -74,7 +78,9 @@ def __init__(self, user_input, temporary_destination, project, remote_destinatio
             all_paths=self.data_list, folder=pathlib.Path(remote_destination or "")
         )
         self.data_list = None
-
+        
+        LOG.debug(self.data)
+
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:83]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 21]
```diff
@@ -141,12 +147,14 @@ def __collect_file_info_local(self, all_paths, folder=pathlib.Path(""), task_nam
                     "overwrite": False,
                     "checksum": "",
                 }
+                LOG.debug(path_key.as_posix())
+                LOG.debug(file_info[path_key.as_posix()])
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:151]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 22]
```diff
@@ -209,16 +217,17 @@ def create_upload_status_dict(self, existing_files, overwrite=False):
...
 
+        LOG.debug(f"status_dict: {status_dict}")
         return status_dict
 
     def check_previous_upload(self, token):
         """Do API call and check for the files in the DB."""
 
         LOG.debug("Checking if files have been previously uploaded.")
-
+        LOG.debug(self.data)
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:227]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 23]
```diff
@@ -227,7 +236,8 @@ def check_previous_upload(self, token):
             json=files,
             error_message="Failed getting information about previously uploaded files",
         )
-
+        LOG.debug(files_in_db)
+
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:240]:
```suggestion
```

File: dds_cli/file_handler_local.py [hunk 24]
```diff
@@ -209,16 +217,17 @@ def create_upload_status_dict(self, existing_files, overwrite=False):
...
     def check_previous_upload(self, token):
         """Do API call and check for the files in the DB."""
 
         LOG.debug("Checking if files have been previously uploaded.")
-
+        LOG.debug(self.data)
         # Get files from db
-        files = list(x for x in self.data)
-
+        # files = list(x for x in self.data)
```
Comment (inaod568) [inline dds_cli/file_handler_local.py:229]:
```suggestion
```

**Comment by inaod568** [comment]
@valyo Ignore the linting and that it's not signed atm but other than that ready

**Comment by valyo** [comment]
> @valyo Ignore the linting and that it's not signed atm but other than that ready

so that one is ready for review, isn't it? and  I should ignore the other one

**Comment by inaod568** [comment]
> > @valyo Ignore the linting and that it's not signed atm but other than that ready
> 
> so that one is ready for review, isn't it? and I should ignore the other one

This one is ready for review, yes 👍🏻 Ignore the other one

**Comment by inaod568** [comment]
> > > @valyo Ignore the linting and that it's not signed atm but other than that ready
> > 
> > 
> > so that one is ready for review, isn't it? and I should ignore the other one
> 
> This one is ready for review, yes 👍🏻 Ignore the other one

@valyo I have now closed the other one so it's clearer

**Review by valyo — APPROVED** [review]
list (incl. with--tree), download, and upload all work fine. 
Skimmed through the code (not super thorough) and didn't see anything strange (not a big surprise :) )

File: dds_cli/utils.py [hunk 25]
```diff
@@ -165,6 +165,22 @@ def perform_request(
...
+        if isinstance(json_input, typing.Dict):
+            for x, y in json_input.items():
+                if isinstance(y, pathlib.Path):
+                    json_input[x] = y.as_posix()
+        elif isinstance(json_input, typing.List):
+            json_input = [x.as_posix() if isinstance(x, pathlib.Path) else x for x in json_input]
+        return json_input
+
+    json = transform_paths(json_input=json)
+    LOG.debug(f"transformed: {json}")
```
Comment (inaod568) [inline dds_cli/utils.py:183]:
```suggestion
```

File: dds_cli/utils.py [hunk 26]
```diff
@@ -165,6 +165,21 @@ def perform_request(
...
+        if isinstance(json_input, typing.Dict):
+            for x, y in json_input.items():
+                if isinstance(y, pathlib.Path):
+                    json_input[x] = y.as_posix()
+        elif isinstance(json_input, typing.List):
+            json_input = [x.as_posix() if isinstance(x, pathlib.Path) else x for x in json_input]
+        return json_input
+
+    json = transform_paths(json_input=json)
     """Perform get request."""
```
Comment (inaod568) [inline dds_cli/utils.py:183]:
```suggestion
    """Perform request."""
```

File: dds_cli/utils.py [hunk 27]
```diff
@@ -165,7 +165,22 @@ def perform_request(
...
+        if isinstance(json_input, typing.Dict):
+            for x, y in json_input.items():
+                if isinstance(y, pathlib.Path):
+                    json_input[x] = y.as_posix()
+        elif isinstance(json_input, typing.List):
+            json_input = [x.as_posix() if isinstance(x, pathlib.Path) else x for x in json_input]
+        return json_input
+
+    json = transform_paths(json_input=json)
+    """Perform request."""
```
Comment (inaod568) [inline dds_cli/utils.py:183]:
```suggestion
    # Perform request.
```

**Comment by inaod568** [comment]
Noticed that I had forgotten to check the source path file option and moved a couple of imports. Otherwise no changes after approval.

**Pull request merged**